### ISSUE TITLE
Feature Request: μP (Maximal Update Parameterization)

### ISSUE THREAD

**Issue by sbhavani** [NVIDIA/Megatron-LM#2824]
## Summary
Request to add support for [μP (Maximal Update Parameterization)](https://arxiv.org/abs/2203.03466) to enable hyperparameter transfer across model scales.

## Motivation
μP allows tuning hyperparameters on small models and transferring them to larger models, reducing the cost of large-scale pretraining. It has been adopted in production models like [Falcon-H1](https://falcon-lm.github.io/blog/falcon-h1/).

## Requested Features
1. **Per-layer Learning Rate Scaling** - Different LR multipliers for specific layers
2. **Width-dependent Initialization** - μP-style init scaling based on model width

## References
- [MuP Paper](https://arxiv.org/abs/2203.03466)
- [Falcon-H1](https://falcon-lm.github.io/blog/falcon-h1/)

**Comment by janEbert** [comment]
Other references to consider for automatic hyperparameter transfer:
- [Modular norm](https://arxiv.org/abs/2405.14813)
- [Modular norm 2](https://arxiv.org/abs/2409.20325)
- [Modular norm 3](https://arxiv.org/abs/2410.21265)
- [Scion optimizer (Muon generalization)](https://arxiv.org/abs/2502.07529)

**Comment by BoxiangW** [comment]
Link to the older issue https://github.com/NVIDIA/Megatron-LM/issues/1263

**Comment by Skylion007** [comment]
Ditto: would be great to add. We have an old fork here that tries to add it, but would be good to expand and properly integrate/test it: https://github.com/LLM360/Megatron-LM-muP

**Comment by Skylion007** [comment]
> Other references to consider for automatic hyperparameter transfer:
> 
>     * [Modular norm](https://arxiv.org/abs/2405.14813)
> 
>     * [Modular norm 2](https://arxiv.org/abs/2409.20325)
> 
>     * [Scion optimizer (Muon generalization)](https://arxiv.org/abs/2502.07529)

Also two other more recent references for parameter transfer:
https://arxiv.org/abs/2505.01618
https://arxiv.org/abs/2512.22382

**Comment by plugyawn** [comment]
Added a PR to address this. The HP transfer is tested on a small dataset, 4-layer transformer, on an A100. Plots and experiment details attached.

**Comment by BoxiangW** [comment]
Help merging https://github.com/NVIDIA/Megatron-LM/pull/3058
